### DISCUSSION TITLE
Custom Queries: How to select rows and export data in .csv

### DISCUSSION THREAD

**Discussion by caitlinnortje** [Budibase/budibase#7517] [Help] [answered]
Hi there,

I'm new to Budibase and trying to figure out the best way to build what I need. 

I have a MySQL Database with related tables: Clients and Contracts (One-to-Many relationship).
I want to display all columns from both the client and contract tables in one table. The user for this app will need to see the data this way.
Currently, I can do this by creating a custom query that JOINs the tables together and use the results in a DataProvider to display all the columns.
However, if I use a custom query I cannot use the select row functionality on the table component and therefore cannot use the export data functionality. 

Any suggestions about how to have all the columns from both tables in one table view with the ability to select single and multiple rows to perform actions such as the export data functionality?

EDIT: Alternatively, is there a way to use the Update State on a button to push all the rows of the table into the Selected Rows array that the Export Data action then uses?

**Comment by melohagan** [comment, marked as answer]
Hi @caitlinnortje 

Instead of using a custom query, you could try getting related fields by using Formulas: https://docs.budibase.com/docs/formula#fields-from-relationships

The export should work for that.

![Screenshot 2022-09-01 at 11 32 06](https://user-images.githubusercontent.com/101575380/187893814-f918a9e2-ca6c-46d7-99ab-2d6851ebb14d.png)

![Screenshot 2022-09-01 at 11 34 54](https://user-images.githubusercontent.com/101575380/187894279-94607271-13e7-4a2d-89d0-122504b078fc.png)

![Screenshot 2022-09-01 at 11 35 20](https://user-images.githubusercontent.com/101575380/187894336-1aa389d9-68fb-4bd2-82ad-04962876f8ac.png)

![Screenshot 2022-09-01 at 11 36 08](https://user-images.githubusercontent.com/101575380/187894478-71aac6d9-250d-4d8a-8f20-c61794fac1bb.png)

<img width="853" alt="Screenshot 2022-09-01 at 11 36 40" src="https://user-images.githubusercontent.com/101575380/187894572-d282a721-a7eb-418c-9786-53a62a3c10cd.png">

**Reply by caitlinnortje** [reply]
Hi @melohagan 

Thank you so much for your response. Thank you for bringing the Formulas to my attention, I did not realise they could be used in this way, very helpful. However, upon further investigation, due to the number of tables, relationships and columns, using Formulas in this way would be too time consuming.

After much investigation and consideration, ideally, I would like to use custom queries to create the views my users require. Is there a way to create my own custom select function for my tables through using checkboxes in my tables or something similar?
My users need to be able to view the data and then do single and multi selects on the data to manipulate it.

**Reply by melohagan** [reply]
Unfortunately I can think of no way to tie custom queries to the **Export Data** action; as you pointed out the row selection does not work for custom queries. 

A user did come up with a custom solution for generating a pdf of the screen, but this may not be what you're after. 
https://github.com/Budibase/budibase/discussions/7278

With that said, using that embed technique it may be possible to use custom code to export the selected rows on screen somehow, but I don't have an easy solution for you!

**Reply by caitlinnortje** [reply]
Ok, thank you for the info!

When doing the Formulas, I have run into a slight complication... I have 3 tables, TableA, TableB and TableC... TableA is related to TableB and TableB is related to TableC.. is there a way to add non-primary key columns from TableA in TableC using formulas?

**Reply by melohagan** [reply]
No it would only allow you to get fields from the immediately related table. If you're wanting to display data from three different tables, is there a chance that you could split some of that into a different screen?

You could create a modal and pass an id to be filtered on via URL variable: 
 - https://docs.budibase.com/docs/url-parameters
 - https://docs.budibase.com/docs/screens#modal-screen

**Comment by JayP718** [comment]
@caitlinnortje 

We had a similar issue. We found a hacky work around but it might suffice for your needs. 

We used a stored procedure to generate a table which can be overwritten anytime a button within Budibase is pressed. It won't work for many users but for our use case it works. We have a multi step form where the first button overwrites the table, then we use that mssql table within Budibase so that way row selection and export data works. Hope this helps.

**Reply by caitlinnortje** [reply]
Hey Jay, thank you so much for sharing... can you explain more about what a procedure is in Budibase? 

Sounds similar to the idea I had of trying to create a View within my MySQL database and using that table in Budibase, however this idea failed because a primary key is needed for the Budibase fetch table function to work and MySQL views don't have a primary key.

**Reply by JayP718** [reply]
https://www.sqlshack.com/learn-mysql-the-basics-of-mysql-stored-procedures/

Absolutely, here's a link which goes into detail but essentially a stored procedure is a set of steps that are executed within a database. With this capability you can tell your database to essentially create the table from the view.

Here's how to do it in mysql.
https://dev.mysql.com/doc/refman/8.0/en/create-table-select.html

After that in the stored procedure you can set a primary key from the table that was just created.
https://www.w3schools.com/sql/sql_primarykey.ASP

Then within Budibase you can just have a Budibase "table" component which refers to this mysql table. Just keep in mind that the data of this table will only be refreshed when the button which runs the stored procedure is clicked within Budibase. So if it's data that is rapidly changing then this might not be the best solution.

**Reply by caitlinnortje** [reply]
Thank you so much. 

This sounds like a very close solution for my purposes. I am currently evaluating whether Budibase is the right choice for building a tool. The tool would be primarily used to view/create/update the data stored in the db as well as download a snapshot of the data for other external uses. This means the users would expect the data to auto update as they use the application therefore it is not ideal if the data only updates when a button is pressed on the 'backend'. Selecting and bulk editing rows is another feature this tool would require and from what I understand, this not possible in Budibase either. It's a shame, because I was so excited about Budibase and it otherwise seems like a great platform, maybe I am too early to adopt it.

**Reply by JayP718** [reply]
@caitlinnortje

Understood, one of the great things about Budibase is they have a great community and are very receptive of feedback. I'd recommend adding a +1 at a feature request I created [here](https://github.com/Budibase/budibase/issues/7348) so perhaps they'll consider adding this ability sooner.

**Reply by caitlinnortje** [reply]
@JayP718 I agree, the community has been great, so thank you for your part in it. This is the first time I have ever posted a question on a discussion board before so it has been a great first time experience. I would love to up vote the feature request, however it is not obvious to me how to do so?

**Reply by melohagan** [reply]
Thanks for the nice words! If you go to that issue Jay posted then you can give it a thumbs up. It doesn't necessarily mean we'll prioritise it, but if a feature request has a lot of up votes then it is some indication that there's high demand!
